sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt

```diff
@@ -439,5 +455,57 @@ internal class ViewAnnotationManagerImpl(
     internal const val EXCEPTION_TEXT_GEOMETRY_IS_NULL = "Geometry can not be null!"
     internal const val EXCEPTION_TEXT_ASSOCIATED_FEATURE_ID_ALREADY_EXISTS =
       "View annotation with associatedFeatureId=%s already exists!"
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun needToReorderZ(
+      positionDescriptorCurrentList: List<ViewAnnotationPositionDescriptor>,
+      positionDescriptorUpdatedList: List<ViewAnnotationPositionDescriptor>
+    ): Boolean {
+      when {
+        // new annotations will be added to layout and trigger on measure anyway
+        positionDescriptorCurrentList.size < positionDescriptorUpdatedList.size -> {
+          return true
+        }
+        positionDescriptorCurrentList.isEmpty() || positionDescriptorUpdatedList.isEmpty() -> {
+          return false
+        }
+        else -> {
+          var currentIndex = 0
+          var updatedIndex = 0
+          // descriptor of items removed from the new list
+          val removedDescriptors = mutableSetOf<String>()
+          while (currentIndex < positionDescriptorCurrentList.size
+            && updatedIndex < positionDescriptorUpdatedList.size) {
+            // skip equal items
+            if (positionDescriptorCurrentList[currentIndex] == positionDescriptorUpdatedList[updatedIndex]) {
+              currentIndex++
+              updatedIndex++
+              continue
+            }
+
+            // if removed descriptors list contains the item from the new list - it means
+            // the element was not removed but moved, need to invalidate Z order
+            if (removedDescriptors.contains(positionDescriptorUpdatedList[currentIndex].identifier)) {
```

Comment:
```suggestion
            if (removedDescriptors.contains(positionDescriptorUpdatedList[updatedIndex].identifier)) {
```

Reply:
Nice catch! Added a test case that crashes here also.